docs/src/design/chiplets/memory.md

```diff
@@ -197,110 +195,54 @@ The amortized cost of reading or writing a single value is between $4$ and $5$ t
-$$
-
-where $\alpha_0$, $\alpha_1$, $\alpha_2$, etc. are random values sent from the verifier to the prover for use in permutation checks.
-
-Note that the values from the top of the stack are added into the row twice: once for "old" values and once for "new" values. We can do this because old and new values in the memory table row corresponding the load operation are the same.
-
-#### Writing to memory
-
-`MSTORE` operation is used to move values from the top of the stack to the memory. This operation works as follows:
-
-1. Pop the top element from the stack and interpret it as memory address.
-2. Perform a lookup into the memory table at the specified address and using the current values of context and clock cycle registers. This creates a row in the lookup table corresponding to the store operation.
 
-Graphically, this looks like so:
+Where $\Delta i = i' - i - 1$.
 
-![memory_writing_to_memory](../../assets/design/chiplets/memory/memory_writing_to_memory.png)
-
-Note that as a result of this operation the stack is shifted to the left by one, and the values saved to memory remain on the stack.
+In addition to this constraint, we also need to make sure that values in registers $d_0$ and $d_1$ are less than $2^{16}$, and this can be done with permutation-based range checks.
 
-Denoting stack registers as $s_j$, helper registers as $h_j$, clock cycle register as $i$, and context register as $c$, we can compute the lookup row value as follows:
+Next, we need to make sure that values at a given memory address are always initialized to $0$. This can be done with the following constraint:
 
-$$
-v = \alpha_0 + \alpha_1 \cdot c + \alpha_2 \cdot s_0 + \alpha_3 \cdot i + \sum_{j=0}^3(\alpha_{j+4} \cdot h_j) + \sum_{j=0}^3(\alpha_{j+8} \cdot s_j')
+>$$
+(n_0 + (1 - n_0) \cdot n_1) \cdot u_i' = 0 \text{ for } i \in \{0, 1, 2, 3\} \text{ | degree} = 5
 $$
 
-where $\alpha_0$, $\alpha_1$, $\alpha_2$ etc... are random values sent from the verifier to the prover for use in permutation checks. Values for the helper registers $h_0, ...,  h_3$ are provided by the VM non-deterministically.
+Thus, when either the context changes, or the address changes, values in $u_i$ columns are guaranteed to be zeros.
 
-We also need to make sure that the saved values remained on the stack. This can be done with the following constraint:
+Lastly, we need to make sure that for the same context/address combination, the $v_i$ columns of the current row are equal to the corresponding $u_i$ columns of the next row. This can be done with the following constraints:
 
-$$
-s_i' - s_{i + 1} = 0
+>$$
+(1 - n_0) \cdot (1 - n_1) \cdot (u_i' - v_i) = 0 \text{ for } i \in \{0, 1, 2, 3\} \text{ | degree} = 5
 $$
 
-where $i \in \{0, 1, 2, 3\}$.
+On the stack side, for every memory access request, a corresponding value is divided out of the $b_{chip}$ column. Specifics of how this is done are described [here](../stack/io_ops.md#memory-access-operations).
```

Comment:
This new constraint section description is an improvement overall, but I think this last bit is confusing, since we haven't provided the "corresponding value" here and specified that it needs to be multiplied into $b_{chip}$ on this side. We've only provided the constraints required for keeping the memory chiplet internally consistent. I think it's important to include information about the lookup here as well, with the calculation of the memory row value. For people trying to understand how these things are connected, I think seeing the formula and a brief reference to it in both places with links to the corresponding section will be helpful (at least, it would have been for me)

Reply:
Ah - I've actually written a whole sub-section about this - but must have been over eager with deleting things and it didn't get committed. I've restored it now.